server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -137,13 +139,31 @@ void sendResponse(
             compressionScheme
         );
         response.mustIncRef();
-        sendMessage(channel, message, responseStatsConsumer, () -> {
+        try {
+            sendMessage(channel, message, responseStatsConsumer, () -> {
+                try {
+                    messageListener.onResponseSent(requestId, action, response);
```

Comment:
Ah I see, we call this on the exception paths too. It seems wrong to call `onResponseSent` if we didn't actually send the response, but that's how it works today. We can clean this up later.